sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs

```diff
@@ -220,6 +263,121 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
+                n => bail!(TransportError::CommunicationError(format!(
+                    "HyperDebug error: {}",
+                    n
+                ))),
+            }
+            let skip_bytes = resp.struct_size as usize - (GPIO_MONITORIN_HEADER_SIZE - 1);
+
+            let mut databytes: Vec<u8> = Vec::new();
+            databytes.extend_from_slice(&resp.data[..bytecount - GPIO_MONITORIN_HEADER_SIZE]);
+
+            while databytes.len() < skip_bytes + resp.transcript_size as usize {
+                let original_length = databytes.len();
+                databytes.resize(original_length + 64, 0u8);
+                let c = self.inner.usb_device.borrow().read_bulk(
+                    cmsis_interface.in_endpoint,
+                    &mut databytes[original_length..],
+                )?;
+                databytes.resize(original_length + c, 0u8);
+            }
+
+            // Figure out how many of the low bits are used for storing the index of the signal
+            // hanving changed. (If only one signal, no bits are used, if two signals, then one
+            // bit is used, if three or four, then two bits are used, etc.)
+            let mut signal_bits = 0;
+            while (pin_names.len() - 1) >> signal_bits != 0 {
+                signal_bits += 1;
+            }
+            let signal_mask = (1u64 << signal_bits) - 1;
+
+            let mut cur_time: u64 = resp.start_timestamp;
+            let mut cur_level = Vec::new();
+            for i in 0..pin_names.len() {
+                cur_level.push(resp.start_levels & (1u16 << i) != 0);
+            }
+            let mut events = Vec::new();
+            let mut idx = skip_bytes;
+
+            // Now decode the list of events, each consisting of a variable legnth encoded 64-bit
+            // integer.
+            while idx < databytes.len() {
+                // Read 7 bits from each byte, least significant byte first.  High bit of one
+                // indicates more bytes belong to the same value.
+                let mut value = 0u64;
+                let mut shift = 0;
+                loop {
+                    value |= ((databytes[idx] & 0x7F) as u64) << shift;
+                    shift += 7;
+                    idx += 1;
+                    if (databytes[idx - 1] & 0x80) == 0 {
+                        break;
+                    }
+                }
+
+                // The 64-bit value consists of two parts, the lower `signal_bits` bits indicate
+                // which signal had an edge, the upper bits indicate the number of microseconds
+                // since the previous event (on any signal, not necessarily on that same one).
+                cur_time += value >> signal_bits;
+                events.push(MonitoringEvent {
+                    signal_index: (value & signal_mask) as u8,
+                    edge: if cur_level[(value & signal_mask) as usize] {
+                        cur_level[(value & signal_mask) as usize] = false;
```

Comment:
This could be a `cur_level[index] ^= true;` after the push, or, `cur_level ^= 1 << index;` if you opted to use bitmask instead of creating an array.

Reply:
Done